src/dns.c

```diff
@@ -2403,30 +2430,36 @@ hsk_dns_name_alloc(
 }
 
 bool
-hsk_dns_name_dirty(const char *name) {
-  char *s = (char *)name;
+hsk_dns_verify_name(const char *name) {
+  int len = strlen(name);
+  if (len > HSK_DNS_MAX_LABEL)
+    return false;
 
-  while (*s) {
-    uint8_t c = (uint8_t)*s;
+  char lower[HSK_DNS_MAX_LABEL + 1];
+  strcpy(lower, name);
+  hsk_to_lower(lower);
+
+  for (int i = 0; i < len; i++) {
+    int c = (uint8_t)lower[i];
 
     switch (c) {
-      case 0x28 /*(*/:
-      case 0x29 /*)*/:
-      case 0x3b /*;*/:
-      case 0x20 /* */:
-      case 0x40 /*@*/:
-      case 0x22 /*"*/:
-      case 0x5c /*\\*/:
-        return true;
+      case 0x5f: /* _ */
+      case 0x2d: /* - */
+        if (i == 0 || i == len - 1) {
+          return false;
+        } else {
+          continue;
+        }
     }
 
-    if (c < 0x20 || c > 0x7e)
-      return true;
-
-    s += 1;
+    if (c < 0x30 ||
```

Comment:
You could remove `hsk_to_lower` and `lower` variable totally.. E.g:
```c
    // upper to lower case
    if (c >= 0x41 && c <= 0x5a)
      c ^= 0x20;
```

c ^= 0x20 will switch from upper to lower and lower to upper :)